WordPress/Classes/Services/PostCoordinator.swift

```diff
@@ -23,18 +25,27 @@ class PostCoordinator: NSObject {
         let mediaCoordinator = MediaCoordinator.shared
 
         if mediaCoordinator.isUploadingMedia(for: post) {
-            mediaCoordinator.addObserver({ (media, state) in
+            // Only observe if we're not already
+            guard observerUUIDs[post] == nil else {
+                return
+            }
+
+            let uuid = mediaCoordinator.addObserver({ (media, state) in
                 switch state {
                 case .ended:
                     self.updateReferences(to: media, in: post)
+
                     // Let's check if media uploading is still going, if all finished with success then we can upload the post
                     if !mediaCoordinator.isUploadingMedia(for: post) && !post.hasFailedMedia {
+                        self.removeObserver(for: post)
                         self.upload(post: post)
                     }
                 default:
```

Comment:
We need to add a case where media the last media for the posts ends up by failing. Something like this:
```               
case  .failed(_):
                    if !mediaCoordinator.isUploadingMedia(for: post) {
                        self.removeObserver(for:post)                    
                    }
```

Reply:
We don't want to remove the observer in the case where all of the media has failed – in this situation, the media coordinator will display a Notice to the user, giving them the chance to retry the uploads. If we leave the observer in place here, then the user can retry the uploads and the observer block will pick it up and upload the post afterwards.

Reply:
Humm fair point, let's go forward like this for now, I will check on my further work for post upload cancel if all is good.

Reply:
I think you're right, there is a situation where the observer could be left hanging, but I'm not sure the right point to detect that and remove it. Here's one scenario:

* A user starts publishing a post that has media to upload. We attach an observer.
* The media fails, but the user chooses not to restart it. They go into the post and remove the media.
* They then try again to publish the post. The post publishes successfully, but we never removed the observer.